### PR TITLE
RATIS-2116. Fix the issue where RaftServerImpl.appendEntries may be blocked indefinitely

### PR DESCRIPTION
**Pull request by sunhaibotb**
## What changes were proposed in this pull request?

This pull request fixes the issue where RaftServerImpl.appendEntries may be blocked indefinitely.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-2116

## How was this patch tested?

unit tests

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java [hunk 1]
```diff
@@ -356,7 +361,9 @@ private boolean shouldFlush() {
     } else if (pendingFlushNum >= forceSyncNum) {
       return true;
     }
-    return pendingFlushNum > 0 && queue.isEmpty();
+
+    final Task nextTask = queue.peek();
+    return pendingFlushNum > 0 && (nextTask == null || nextTask.getType() != TaskType.WRITE_LOG);
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java:366]: Let's simply use `instanceof`.
```java
    final Task peeked = queue.peek();
    // when the queue is empty, peek() returns null and (null instanceof WriteLog) is false
    return pendingFlushNum > 0 && !(peeked instanceof WriteLog);
```

**Review by szetszwo — COMMENTED** [review]
@sunhaibotb , thanks a lot for working on this!  We may use `instanceof` in this case instead of adding an `enum`; see the comment inlined.  Btw, thanks for adding a new test.

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java (hunk 1 above)
Comment (sunhaibotb) [reply ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java]: Thank for the review, @szetszwo . I've simplified it to use instanceof now. Please review again.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**